dvc/command/completion.py

```diff
@@ -0,0 +1,203 @@
+import argparse
+import io
+import logging
+import sys
+
+from dvc.command.base import CmdBase, append_doc_link
+
+logger = logging.getLogger(__name__)
+GLOBAL_OPTIONS = ["-h", "--help", "-q", "--quiet", "-v", "--verbose"]
+ROOT_PREFIX = "_dvc"
+UNCOMPLETABLE_POSITIONALS = {
+    "rev",
+    "url",
+    "args",
+    "name",
+    "option",
+    "value",
+    "command",
+}
+
+
+def print_bash_commands(parser, prefix=ROOT_PREFIX, fd=None):
+    """Recursive parser traversal, printing bash helper syntax."""
+    positionals = parser._get_positional_actions()
+    commands = []
+
+    if prefix == ROOT_PREFIX:  # skip root options
+        pass
+    else:
+        opts = [
+            opt for sub in positionals if sub.choices for opt in sub.choices
+        ]
+        opts += sum(
+            (opt.option_strings for opt in parser._get_optional_actions()), []
+        )
+        # use list rather than set to maintain order
+        opts = [i for i in opts if i not in GLOBAL_OPTIONS]
+        opts = " ".join(opts)
+        print(f"{prefix}='{opts}'", file=fd)
+
+    dest = []
+    for sub in positionals:
+        if sub.choices:
+            for cmd in sorted(sub.choices):
+                commands.append(cmd)
+                print_bash_commands(
+                    sub.choices[cmd], f"{prefix}_{cmd.replace('-', '_')}", fd
+                )
+        elif not any(i in sub.dest for i in UNCOMPLETABLE_POSITIONALS):
+            dest.append(sub.dest)
+        else:
+            logger.debug(f"uncompletable:{prefix}:{dest}")
+    if dest:
+        if not {"targets", "target"}.intersection(dest):
```

Comment:
I would consider adding exclusion for particular commands. As of now, we add `_dvc_compgen_files` for `get`, `get-url` and `import` too, which I think does not make sense.

Reply:
i think it does make sense in terms of completing output directory path but that should be sorted out in a separate PR. Don't have a strong opinion on whether or not they should be excluded for now.

Reply:
Right, that makes sense. I was too much focused on the fact that, when importing, ideally you would get autocomplete for target repo.

Reply:
outdated - removed. Also fixed (for `zsh` at least)